src/rust/lib/flexer/src/group.rs

```diff
@@ -0,0 +1,256 @@
+use crate::automata::pattern::Pattern;
+use crate::automata::nfa::NFA;
+use crate::group::rule::Rule;
+
+use itertools::Itertools;
+
+pub mod rule;
+
+/// Group of a set of rules.
+struct Group {
+    /// ID of group.
+    pub id     : usize,
+    /// Name of group.
+    pub name   : String,
+    /// Parent of group.
+    pub parent : Option<Box<Group>>,
+    /// Set of rules. See Rule for more information.
+    pub rules  : Vec<Rule>,
+    /// Function that is called on exiting group.
+    pub finish : Box<dyn FnMut()>,
+}
+
+impl Default for Group {
+    fn default() -> Self {
+        Group {
+            name   : Default::default(),
+            id     : Default::default(),
+            parent : Default::default(),
+            rules  : Default::default(),
+            finish : Box::new(||{}),
+        }
+    }
+}
+
+impl Group {
+
+    /// Adds new rule to group.
+    pub fn add_rule(&mut self, rule:Rule) {
+        self.rules.push(rule)
+    }
+
+    /// Returns rule builder for given pattern.
+    pub fn rule(&mut self, pattern:Pattern) -> rule::Builder<impl FnMut(Rule) + '_> {
+        rule::Builder { pattern, callback: move |rule| self.add_rule(rule) }
+    }
+
+    /// All rules including parent rules.
+    pub fn rules(&self) -> Vec<&Rule> {
+        let mut parent = &self.parent;
+        let mut rules  = (&self.rules).iter().collect_vec();
+        while let Some(state) = parent {
+            rules.extend((&state.rules).iter());
+            parent = &state.parent;
+        }
+        rules
+    }
+
+    /// Canonical name of given rule.
+    pub fn rule_name(&self, rule_ix:usize) -> String {
```

Comment:
what is `rule_ix` and why it is `usize`? Sohuld be newtype

Reply:
It isn't newtype on purpose (see how it is used in scala). It shouldn't be public though.